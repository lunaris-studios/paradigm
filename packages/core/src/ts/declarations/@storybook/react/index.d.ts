import * as Abstract from "@paradigmjs/abstract";
import * as Addons from "@storybook/addons";
import * as SR from "@storybook/react";

import "@storybook/react";

declare module "@storybook/react" {
	/**
	 * Metadata to configure the stories for an abstract component.
	 */
	declare type StorybookComponent =
		| React.ComponentType
		| Abstract.AbstractComponent
		| Abstract.AbstractPureComponent;
	export declare type Meta<
		TComponent extends StorybookComponent = any,
		TArgs = Addons.DefaultArgs
	> = Addons.BaseMeta<TComponent> &
		Addons.Annotations<Component.Spinner, React.ReactElement<unknown>>;
}
