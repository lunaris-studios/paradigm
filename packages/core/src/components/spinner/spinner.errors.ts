import * as Common from "~/common";

const ns = Common.DISPLAYNAME_PREFIX;

export const SPINNER_WARN_VALUE_OUT_OF_BOUNDS = `${ns} <Spinner> props.value is out of bounds. Values below 0 or above 1 will be interpreted as 0 or 1 respectively. `;
