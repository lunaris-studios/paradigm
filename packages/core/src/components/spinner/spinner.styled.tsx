import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";
import * as SC from "styled-components";
import * as Util from "@paradigmjs/util";
// re-import `styled-components` development mode DOM classnames.
import styled, { css, keyframes } from "styled-components";

import * as Common from "~/common";

/**
 * [Spinner]
 * - [Spinner.Container(WRAPPER)]
 */

interface Spinner {
	Container: Util.SC.Styled<"div", ISpinnerContainerProps>;

	SVG: SpinnerSVG;
}

export const Spinner = {} as Spinner;

/**
 * [Spinner.Container]
 */

interface ISpinnerContainerProps {}

Spinner.Container = styled("div")`
	align-items: center;
	/* center animation container inside parent element to isolate layout */
	display: flex;
	justify-content: center;

	/* allow paths to overflow container -- critical for edges of circles! */
	overflow: visible;
	vertical-align: middle;
`;

/**
 * [Spinner.SVG]
 * - [Spinner.SVG.Container(WRAPPER)]
 */

interface SpinnerSVG {
	Container: Util.SC.Styled<"svg", ISpinnerSVGContainerProps>;

	Path: SpinnerSVGPath;
}

Spinner.SVG = {} as SpinnerSVG;

/**
 * [Spinner.SVG.Container]
 */

interface ISpinnerSVGContainerProps extends Common.IIntentProps {
	isSpinning: boolean;
}

const spinnerAnimation = keyframes`{
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}`;

Spinner.SVG.Container = styled("svg")`
	display: block;
	fill: ${Protocol.Color.TRANSPARENT};
	animation: ${(props: ISpinnerSVGContainerProps) =>
			props.isSpinning ? spinnerAnimation : "none"}
		500ms linear infinite;
`;

/**
 * [Spinner.SVG.Path]
 * - [Spinner.SVG.Path.Shared(CSS)]
 * - [Spinner.SVG.Path.Track(ELEMENT)]
 * - [Spinner.SVG.Path.Head(ELEMENT)]
 */

interface SpinnerSVGPath {
	Shared: SC.FlattenSimpleInterpolation;

	Head: Util.SC.Styled<"path", ISpinnerSVGPathHeadProps>;
	Track: Util.SC.Styled<"path", ISpinnerSVGPathTrackProps>;
}

Spinner.SVG.Path = {} as SpinnerSVGPath;

/**
 * [Spinner.SVG.Path.Shared]
 */

Spinner.SVG.Path.Shared = css`
	fill-opacity: 0;
`;

/**
 * [Spinner.SVG.Path.Head]
 */

interface ISpinnerSVGPathHeadProps extends Common.IIntentProps {}

const SPINNER_SVG_PATH_HEAD_HEAD = Protocol.bind("scheme", {
	[Protocol.Scheme.DARK]: Protocol.Color.BLACK_3,
	[Protocol.Scheme.LIGHT]: Protocol.Color.BLACK_5,
});

// TODO (sam): Add intent / color injection
Spinner.SVG.Path.Head = styled("path")`
	${Spinner.SVG.Path.Shared}
	/*  */
	stroke: ${SPINNER_SVG_PATH_HEAD_HEAD};
	stroke-linecap: round;
	transform-origin: center;
	transition: stroke-dashoffset 200ms ${Protocol.TRANSITION_EASE};
`;

/**
 * [Spinner.SVG.Path.Track]
 */

interface ISpinnerSVGPathTrackProps extends Common.IIntentProps {}

const SPINNER_SVG_PATH_HEAD_TRACK = Protocol.bind("scheme", {
	[Protocol.Scheme.DARK]: Protocol.Color.BLACK_2,
	[Protocol.Scheme.LIGHT]: Protocol.Color.BLACK_4,
});

Spinner.SVG.Path.Track = styled("path")`
	${Spinner.SVG.Path.Shared}
	/*  */
	stroke: ${SPINNER_SVG_PATH_HEAD_TRACK};
`;
