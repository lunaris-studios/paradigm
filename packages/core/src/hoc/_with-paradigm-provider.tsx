import * as Abstract from "@paradigmjs/abstract";
import * as React from "react";

import * as BIN from "~/bin";
import * as Util from "~/util";

/**
 * A Higher-Order-Component that overrides the existing root
 * <ParadigmProvider> instance. ALlows the user to override
 * default `theme.binds` to test specific use-cases.
 * @param defaultBinds Explicit binds to override defaults
 * @param defaultProps Prop object to pass to the target component
 */
export function withParadigmProvider<
	TParadigmThemeBinds extends Partial<BIN.IParadigmProviderProps["theme"]["binds"]>
>(
	defaultBinds: TParadigmThemeBinds = BIN.ParadigmProvider.defaultProps.theme
		.binds as TParadigmThemeBinds,
) {
	return function WrappedTarget<TTargetComponentProps extends object>(
		TargetComponent: React.ComponentType<TTargetComponentProps>,
	): React.ComponentClass<Partial<TTargetComponentProps>> {
		return class WithParadigmProvider extends React.PureComponent<
			Partial<TTargetComponentProps>
		> {
			public static readonly displayName = `WithParadigmProvider(${TargetComponent.displayName})`;

			public render(): JSX.Element {
				const paradigmTheme = Object.freeze<BIN.IParadigmProviderProps["theme"]>({
					...BIN.ParadigmProvider.defaultProps.theme,
					binds: {
						...BIN.ParadigmProvider.defaultProps.theme.binds,
						...defaultBinds,
					},
				});

				const combinedTargetProps = Object.freeze({
					...this.props,
					...TargetComponent.defaultProps,
				});

				return (
					<BIN.ParadigmProvider theme={paradigmTheme}>
						<TargetComponent {...(combinedTargetProps as TTargetComponentProps)} />
					</BIN.ParadigmProvider>
				);
			}
		};
	};
}
