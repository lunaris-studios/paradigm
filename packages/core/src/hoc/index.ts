export * from "./_with-paradigm-provider";
