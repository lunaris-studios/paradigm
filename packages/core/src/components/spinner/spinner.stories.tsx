import * as Abstract from "@paradigmjs/abstract";
import * as React from "react";
import * as SR from "@storybook/react";

import * as Component from "./spinner";

const Template: SR.Story<Component.ISpinnerProps> = (args) => (
	<Component.Spinner {...args} />
);

export const Basic = Template.bind({});
Basic.args = {};

export default {
	title: "Spinner",
	component: Component.Spinner,
} as SR.Meta;
