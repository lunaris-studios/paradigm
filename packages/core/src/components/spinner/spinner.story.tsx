import * as React from "react";
import * as Knobs from "@storybook/addon-knobs";
import * as Actions from "@storybook/addon-actions";

import { Spinner } from "./spinner";

export default {
	title: "Spinner",
	component: Spinner,
	decorators: [Knobs.withKnobs],
};

export const interactive = () => {
	return <Spinner />;
};
interactive.story = {
	parameters: {
		jest: ["spinner.test.tsx"],
	},
};
