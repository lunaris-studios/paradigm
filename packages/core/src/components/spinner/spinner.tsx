import * as React from "react";

import * as Common from "~/common";
import * as Components from "~/components";
import * as Util from "~/util";

import * as Styled from "./spinner.styled";
import * as Errors from "./spinner.errors";

// see http://stackoverflow.com/a/18473154/3124288 for calculating arc path
const R = 45;
const SPINNER_TRACK = `M 50,50 m 0,-${R} a ${R},${R} 0 1 1 0,${
	R * 2
} a ${R},${R} 0 1 1 0,-${R * 2}`;

// unitless total length of SVG path, to which stroke-dash* properties are relative.
// https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/pathLength
// this value is the result of `<path d={SPINNER_TRACK} />.getTotalLength()` and works in all browsers:
const PATH_LENGTH = 280;

const MIN_SIZE = 10;
const STROKE_WIDTH = 4;
const MIN_STROKE_WIDTH = 16;

export enum SpinnerSize {
	SIZE_SMALL = 20,
	SIZE_STANDARD = 50,
	SIZE_LARGE = 100,
}

export interface ISpinnerProps extends Common.IProps, Common.IIntentProps {
	/**
	 * Width and height of the spinner in pixels. The size cannot be less than
	 * 10px.
	 *
	 * Constants are available for common sizes:
	 * - `Spinner.SIZE_SMALL = 20px`
	 * - `Spinner.SIZE_STANDARD = 50px`
	 * - `Spinner.SIZE_LARGE = 100px`
	 *
	 * @default Spinner.SIZE_STANDARD = 50
	 */
	size: number;

	/**
	 * HTML tag for the two wrapper elements. If rendering a `<Spinner>` inside
	 * an `<svg>`, change this to an SVG element like `"g"`.
	 * @default "div"
	 */
	tagName: keyof JSX.IntrinsicElements;

	/**
	 * A value between 0 and 1 (inclusive) representing how far along the operation is.
	 * Values below 0 or above 1 will be interpreted as 0 or 1 respectively.
	 * Omitting this prop will result in an "indeterminate" spinner where the head spins indefinitely.
	 */
	value: number;
}

const defaultProps = Object.freeze<ISpinnerProps>({
	size: SpinnerSize.SIZE_STANDARD,
	tagName: "div",
	value: 0.25,
});

export class Spinner extends Components.AbstractPureComponent<ISpinnerProps, {}> {
	public static displayName = `${Common.DISPLAYNAME_PREFIX}.Spinner`;

	static readonly defaultProps: ISpinnerProps = defaultProps;

	public static Styled = Styled;
	public static Errors = Errors;

	public static readonly SIZE_SMALL = SpinnerSize.SIZE_SMALL;
	public static readonly SIZE_STANDARD = SpinnerSize.SIZE_STANDARD;
	public static readonly SIZE_LARGE = SpinnerSize.SIZE_LARGE;

	public componentDidUpdate(prevProps: ISpinnerProps) {
		if (prevProps.value !== this.props.value) {
			// IE/Edge: re-render after changing value to force SVG update
			this.forceUpdate();
		}
	}

	public render() {
		const { intent, value, tagName } = this.props;
		const size = this.getSize();

		// keep spinner track width consistent at all sizes (down to about 10px).
		const strokeWidth = Math.min(
			MIN_STROKE_WIDTH,
			(STROKE_WIDTH * Spinner.SIZE_LARGE) / size,
		);
		const strokeOffset =
			PATH_LENGTH - PATH_LENGTH * (value == null ? 0.25 : Util.clamp(value, 0, 1));

		const isSpinning = Boolean(value != null);

		// multiple DOM elements around SVG are necessary to properly isolate animation:
		// - SVG elements in IE do not support anim/trans so they must be set on a parent HTML element.
		// - SPINNER_ANIMATION isolates svg from parent display and is always centered inside root element.
		return (
			<Styled.Spinner.Container as={tagName}>
				<Styled.Spinner.SVG.Container
					width={size}
					height={size}
					strokeWidth={strokeWidth.toFixed(2)}
					viewBox={this.getViewBox(strokeWidth)}
					intent={intent}
					isSpinning={isSpinning}
				>
					<Styled.Spinner.SVG.Path.Head
						d={SPINNER_TRACK}
						pathLength={PATH_LENGTH}
						strokeDasharray={`${PATH_LENGTH} ${PATH_LENGTH}`}
						strokeDashoffset={strokeOffset}
						intent={intent}
					/>
					<Styled.Spinner.SVG.Path.Track d={SPINNER_TRACK} intent={intent} />
				</Styled.Spinner.SVG.Container>
			</Styled.Spinner.Container>
		);
	}

	protected validateProps(props: ISpinnerProps) {
		const { value } = props;
		if (value > 1 || value < 0) {
			console.warn(Errors.SPINNER_WARN_VALUE_OUT_OF_BOUNDS);
		}
	}

	/**
	 * Resolve size to a pixel value.
	 * Size can be set by props, default, or minimum constant.
	 */
	private getSize() {
		const { size } = this.props;
		if (size == null) {
			return Spinner.SIZE_STANDARD;
		}
		return Math.max(MIN_SIZE, size);
	}

	/** Compute viewbox such that stroked track sits exactly at edge of image frame. */
	private getViewBox(strokeWidth: number) {
		const radius = R + strokeWidth / 2;
		const viewBoxX = (50 - radius).toFixed(2);
		const viewBoxWidth = (radius * 2).toFixed(2);
		return `${viewBoxX} ${viewBoxX} ${viewBoxWidth} ${viewBoxWidth}`;
	}
}
