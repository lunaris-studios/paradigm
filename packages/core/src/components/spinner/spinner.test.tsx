import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";
import * as SCTU from "styled-components/test-utils";
import * as Testing from "@testing-library/react";

import * as HOC from "~/hoc";

import * as Component from "./spinner";
import * as Styled from "./spinner.styled";

describe("Spinner", () => {
	it("renders correctly", () => {
		const node = Testing.render(<Component.Spinner />);

		const SpinnerContainer = SCTU.find(node.baseElement, Styled.Spinner.Container);
		expect(SpinnerContainer).toMatchSnapshot();
	});

	it("clamps underflow `value`", () => {
		/**
		 * Assume a `stroke-dashoffset` equal to the length of the spinner's
		 * total path.
		 */
		const node = Testing.render(<Component.Spinner value={-1} />);
		const SpinnerSVGPathhead = SCTU.find(node.baseElement, Styled.Spinner.SVG.Path.Head);

		expect(SpinnerSVGPathhead).toHaveAttribute(
			"stroke-dashoffset",
			`${Component.PATH_LENGTH}`,
		);
	});

	it("clamps overflow `value`", () => {
		/**
		 * Assume a `stroke-dashoffset` equal to zero.
		 */
		const node = Testing.render(<Component.Spinner value={2} />);
		const SpinnerSVGPathhead = SCTU.find(node.baseElement, Styled.Spinner.SVG.Path.Head);

		expect(SpinnerSVGPathhead).toHaveAttribute("stroke-dashoffset", "0");
	});

	it("<Styled.Spinner.SVG.Path.Head> responds to color scheme change when [Scheme.DARK] is active", () => {
		const Spinner = HOC.withParadigmProvider({ scheme: Protocol.Scheme.DARK })(
			Component.Spinner,
		);
		const node = Testing.render(<Spinner />);

		const SpinnerSVGPathHead = SCTU.find(node.baseElement, Styled.Spinner.SVG.Path.Head);
		expect(SpinnerSVGPathHead).toHaveStyleRule("stroke", Protocol.Color.BLACK_3);
	});

	it("<Styled.Spinner.SVG.Path.Head> responds to color scheme change when [Scheme.LIGHT] is active", () => {
		const Spinner = HOC.withParadigmProvider({ scheme: Protocol.Scheme.LIGHT })(
			Component.Spinner,
		);
		const node = Testing.render(<Spinner />);

		const SpinnerSVGPathHead = SCTU.find(node.baseElement, Styled.Spinner.SVG.Path.Head);
		expect(SpinnerSVGPathHead).toHaveStyleRule("stroke", Protocol.Color.BLACK_5);
	});

	it("<Styled.Spinner.SVG.Path.Track> responds to color scheme change when [Scheme.DARK] is active", () => {
		const Spinner = HOC.withParadigmProvider({ scheme: Protocol.Scheme.DARK })(
			Component.Spinner,
		);
		const node = Testing.render(<Spinner />);

		const SpinnerSVGPathTrack = SCTU.find(
			node.baseElement,
			Styled.Spinner.SVG.Path.Track,
		);
		expect(SpinnerSVGPathTrack).toHaveStyleRule("stroke", Protocol.Color.BLACK_2);
	});

	it("<Styled.Spinner.SVG.Path.Track> responds to color scheme change when [Scheme.LIGHT] is active", () => {
		const Spinner = HOC.withParadigmProvider({ scheme: Protocol.Scheme.LIGHT })(
			Component.Spinner,
		);
		const node = Testing.render(<Spinner />);

		const SpinnerSVGPathTrack = SCTU.find(
			node.baseElement,
			Styled.Spinner.SVG.Path.Track,
		);
		expect(SpinnerSVGPathTrack).toHaveStyleRule("stroke", Protocol.Color.BLACK_4);
	});
});
