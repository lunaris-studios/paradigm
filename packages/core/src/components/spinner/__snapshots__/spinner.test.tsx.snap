// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders correctly 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: visible;
  vertical-align: middle;
}

.c1 {
  display: block;
  fill: rgba(16,22,26,0);
  -webkit-animation: cpltiX 500ms linear infinite;
  animation: cpltiX 500ms linear infinite;
}

.c2 {
  fill-opacity: 0;
  stroke-linecap: round;
  -webkit-transform-origin: center;
  -ms-transform-origin: center;
  transform-origin: center;
  -webkit-transition: stroke-dashoffset 200ms cubic-bezier(0.4,1,0.75,0.9);
  transition: stroke-dashoffset 200ms cubic-bezier(0.4,1,0.75,0.9);
}

.c3 {
  fill-opacity: 0;
}

<div
  class="c0"
>
  <svg
    class="c1"
    height="50"
    stroke-width="8.00"
    viewBox="1.00 1.00 98.00 98.00"
    width="50"
  >
    <path
      class="c2"
      d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
      pathLength="280"
      stroke-dasharray="280 280"
      stroke-dashoffset="210"
    />
    <path
      class="c3"
      d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
    />
  </svg>
</div>
`;
